src/main/java/org/chameleoncloud/api/admin/UserGroupRolesResource.java

```diff
@@ -0,0 +1,320 @@
+	}
+	String groupId = group.getId();
+	Set<String> associatedPolicies = groupScopePolicy.getAssociatedPolicies().stream()
+		.filter(p -> "group".equals(p.getType())).map(Policy::getName).collect(Collectors.toSet());
+	String associatedPolicyName = UserGroupRoles.formatPolicyName(policyName, groupId);
+	if (!associatedPolicies.contains(associatedPolicyName)) {
+	    // policy not associated
+	    // check if the associated policy exists
+	    ResourceServer rs = groupScopePolicy.getResourceServer();
+	    Policy associatedPolicy = policyStore.findByName(associatedPolicyName, rs.getId());
+	    if (associatedPolicy == null) {
+		// create a policy
+		GroupPolicyRepresentation rep = new GroupPolicyRepresentation();
+		rep.setName(associatedPolicyName);
+		rep.addGroup(subGroup.getId(), true);
+		associatedPolicy = policyStore.create(rep, groupScopePolicy.getResourceServer());
+	    }
+	    groupScopePolicy.addAssociatedPolicy(associatedPolicy);
+	}
+    }
+
+    private void grantMemberPermissions(GroupModel group) {
+	grantPermission("view", group, group, UserGroupRoles.MEMBER);
+	grantPermission("view-members", group, group, UserGroupRoles.MEMBER);
+    }
+
+    private void grantManagerPermissions(GroupModel group, GroupModel managerSubGroup) {
+	grantPermission("manage-membership", group, managerSubGroup, UserGroupRoles.MANAGER);
+    }
+
+    private void grantAdminPermissions(GroupModel group, GroupModel adminSubGroup) {
+	grantPermission("manage", group, adminSubGroup, UserGroupRoles.ADMIN);
+	grantPermission("manage-members", group, adminSubGroup, UserGroupRoles.ADMIN);
+	grantPermission("manage-membership", group, adminSubGroup, UserGroupRoles.ADMIN);
+    }
+
+    @GET
+    @Path("{user}")
+    @NoCache
+    @Produces(MediaType.APPLICATION_JSON)
+    public Response getUserGroupRoles(@QueryParam("group") String groupId, @PathParam("user") String userId) {
+
+	UserModel user = session.users().getUserById(userId, realm);
+	if (user == null) {
+	    return Response.status(Response.Status.NOT_FOUND).entity("user not found!").build();
+	}
+
+	Set<GroupModel> userGroups = user.getGroups();
+
+	if (groupId != null) {
+	    GroupModel group = session.realms().getGroupById(groupId, realm);
+	    Set<String> filterGroups = new HashSet<String>();
+	    filterGroups.add(groupId);
+	    for (GroupModel g : group.getSubGroups()) {
+		filterGroups.add(g.getId());
+	    }
+	    userGroups = userGroups.stream().filter(g -> filterGroups.contains(g.getId())).collect(Collectors.toSet());
+	}
+
+	Set<String> userGroupIds = new HashSet<String>();
+	// lazily grant permissions to user groups
+	for (GroupModel userGroup : userGroups) {
+	    String userGroupName = userGroup.getName();
+	    if (isRootGroup(userGroup)) {
+		grantMemberPermissions(userGroup);
+	    } else {
+		if (userGroupName.endsWith(UserGroupRoles.ADMIN_GROUP_SUFFIX)) {
+		    grantAdminPermissions(userGroup.getParent(), userGroup);
```

Comment:
I'm confused, why is this happening on a GET? Is it not enough to have this on the PUT/POST endpoint?

Reply:
this is for the "first-time"/"one-time" set up so that we don't need a one-time script to set up the existing projects in keycloak.